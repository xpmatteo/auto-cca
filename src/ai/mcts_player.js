import { score } from "./score.js";

class TreeNode {
    constructor(game, parent = null) {
        this.game = game;
        this.parent = parent;
        this.score = 0;
        this.visits = 0;
        this.children = [];
    }

    size() {
        if (this.children.length === 0) {
            return 1;
        }
        return 1 + this.children.reduce((acc, child) => acc + child.size(), 0);
    }

    shape() {
        const result = [];
        function traverse(node, level) {
            if (!result[level]) {
                result[level] = 0;
            }
            result[level]++;
            for (const child of node.children) {
                traverse(child, level + 1);
            }
        }
        traverse(this, 0);
        return result;
    }

    toString() {
        const result = [];
        function traverse(node, level) {
            result.push(`${" ".repeat(level)}${node.score}/${node.visits}`);
            for (const child of node.children) {
                traverse(child, level + 1);
            }
        }
        traverse(this, 0);
        return result.join("\n");
    }
}

export class MctsPlayer {
    constructor(args) {
        this.args = args;
    }

    search() {
        const rootNode = new TreeNode(this.args.game);
        for (let i = 0; i < this.args.iterations; i++) {
            let node = this._select(rootNode);
            let score = this._simulate(node.game);
            this._backpropagate(node, score);
        }
        return rootNode;
    }

    _expand(node) {
        const game = node.game;
        const validCommands = game.validCommands();
        validCommands.forEach((command) => {
            const clone = game.clone();
            clone.executeCommand(command);
            const childNode = new TreeNode(clone, node);
            childNode.command = command;
            node.children.push(childNode);
        });
        return node.children[0];
    }

    _backpropagate(node, score) {
        while (node !== null) {
            node.score += score;
            node.visits++;
            node = node.parent;
        }
    }

    _select(node) {
        while (!node.game.isTerminal()) {
            if (node.children.length === 0) {
                return this._expand(node);
            } else {
                node = node.bestChild();
            }
        }
        return node;
    }

    _simulate(game) {
        return score(game, game.currentSide);
    }
}
