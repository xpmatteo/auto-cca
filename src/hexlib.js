"use strict";

// thanks https://www.redblobgames.com/grids/hexagons/#coordinates

class Hex {
    constructor(q, r, s=undefined) {
        if (s === undefined) {
            s = -q - r;
        }
        assertEquals(0, q + r + s, "q + r + s must be 0");
        this.q = q;
        this.r = r;
    }

    get s() { return - this.q - this.r }
};


class Orientation {
    // start_angle is in multiples of 60Â°
    constructor(f0_, f1_, f2_, f3_, b0_, b1_, b2_, b3_, start_angle_) {
        this.f0 = f0_; 
        this.f1 = f1_; 
        this.f2 = f2_; 
        this.f3 = f3_;
        this.b0 = b0_;
        this.b1 = b1_;
        this.b2 = b2_;
        this.b3 = b3_;
        this.start_angle = start_angle_;
    }
};

class Point {
    x; y;
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
}

class Layout {    
    constructor(orientation, size, origin) {
        this.orientation = orientation;
        this.size = size;
        this.origin = origin;
    }
}

function hex_to_pixel(layout, hex) {
    let M = layout.orientation;
    let x = (M.f0 * hex.q + M.f1 * hex.r) * layout.size.x;
    let y = (M.f2 * hex.q + M.f3 * hex.r) * layout.size.y;
    return new Point(x + layout.origin.x, y + layout.origin.y);
}

function pixel_to_hex(layout, p) {
    const M = layout.orientation;
    let pt = new Point((p.x - layout.origin.x) / layout.size.x, 
                        (p.y - layout.origin.y) / layout.size.y);
    let q = M.b0 * pt.x + M.b1 * pt.y;
    let r = M.b2 * pt.x + M.b3 * pt.y;
    return hex_round(q, r, -q - r);
}

function hex_round(fracq, fracr, fracs) {
    let q = int(round(fracq));
    let r = int(round(fracr));
    let s = int(round(fracs));
    let q_diff = abs(q - fracq);
    let r_diff = abs(r - fracr);
    let s_diff = abs(s - fracs);
    if (q_diff > r_diff && q_diff > s_diff) {
        q = -r - s;
    } else if (r_diff > s_diff) {
        r = -q - s;
    } else {
        s = -q - r;
    }
    return new Hex(q, r, s);
}

const sqrt = Math.sqrt;

const layout_pointy
  = new Orientation(sqrt(3.0), sqrt(3.0) / 2.0, 0.0, 3.0 / 2.0,
                sqrt(3.0) / 3.0, -1.0 / 3.0, 0.0, 2.0 / 3.0,
                0.5);


function assertEquals(expected, actual, message = "Assertion failed") {
    if (expected != actual) {
        throw new Error(message + ": expected " + expected + ", got " + actual);
    }
}

function assertDeepEquals(expected, actual, message = "Assertion failed") {
    if (JSON.stringify(expected) != JSON.stringify(actual)) {
        throw new Error(message + ": expected " + expected + ", got " + actual);
    }
}

                
assertEquals(0, 1, "next: print pixel_to_hex for mouse click");

const test_layout = new Layout(layout_pointy, new Point(50, 60), new Point(10, 100));
assertDeepEquals(layout_pointy, test_layout.orientation);
assertDeepEquals(hex_to_pixel(test_layout, new Hex(0, 0, 0)), new Point(10, 100));
assertEquals(  0, new Hex(1,  -1, 0).s);
assertEquals(-30, new Hex(10, 20)   .s);
