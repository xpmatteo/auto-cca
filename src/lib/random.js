
export function randomShuffleArray(array) {
    for (let i = 0; i < array.length; i++) {
        const j = Math.floor(Math.random() * (array.length - i)) + i;
        [array[i], array[j]] = [array[j], array[i]];
    }
}

/**
 * @template T
 * @param {T[]} array
 * @returns {T}
 */
export function randomElement(array) {
    return array[Math.floor(Math.random() * array.length)];
}

const RANDOM_VALUES = [0.1, 0.3, 0.2, 0.7, 0.9, 0.6, 0.4, 0.8];
let nextRandom = 0;
export function fixedRandom() {
    return RANDOM_VALUES[nextRandom++ % 8];
}
