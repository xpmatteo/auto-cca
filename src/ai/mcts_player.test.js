import { MctsPlayer } from "./mcts_player.js";
import * as fs from "fs";
import makeGame from "../model/game.js";
import { MeleeScenario, NullScenario, TwoOnTwoMeleeScenario } from "../model/scenarios.js";
import { CarthaginianHeavyInfantry, RomanLightInfantry } from "../model/units.js";
import { hexOf } from "../lib/hexlib.js";
import { fixedRandom, resetFixedRandom } from "../lib/random.js";

/**
 * @param {string} fileName
 */
function dumpTreeToFile(root, fileName) {
    const logger = fs.createWriteStream(fileName, {
        flags: 'w'
    });
    function write(str) {
        logger.write(str);
        logger.write("\n");
    }
    root.dumpTree(write);
}

const originalRandom = Math.random;
beforeEach(() => {
    resetFixedRandom();
    Math.random = fixedRandom;
});

afterEach(() => {
    Math.random = originalRandom;
});

const player = new MctsPlayer({
    expansionFactor: 2.1415,
    playoutIterations: 20,
    logfunction: () => {},
});

// fails bc I don't know
xtest('2 on 2', () => {
    const game = makeGame(new TwoOnTwoMeleeScenario());

    const root = player.search(game, 2000);

    const rootAsString = root.toString(6);
    expect(rootAsString).toMatchSnapshot();
});

// disabled bc of bug in simulation of battle back after ignored flags
xtest('melee', () => {
    const game = makeGame(new MeleeScenario());

    const root = player.search(game, 2000);

    const rootAsString = root.toString(6);
    expect(rootAsString).toMatchSnapshot();
});

// disabled bc of evasion phase is not yet implemented
xtest('close combat from light to heavy', () => {
    const game = makeGame(new NullScenario());
    game.placeUnit(hexOf(2, 2), new CarthaginianHeavyInfantry());
    game.placeUnit(hexOf(2, 3), new RomanLightInfantry());

    const root = player.search(game, 30);

    const rootAsString = root.toString(7);
    expect(rootAsString).toMatchSnapshot();
});
