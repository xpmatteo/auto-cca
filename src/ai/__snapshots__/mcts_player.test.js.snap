// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MCTS snapshot testing 2 on 2 1`] = `
"526/2000: undefined -> Roman play one card -> 2
 480/1628: PlayCard(Order Heavy Troops) -> Roman order 3 units -> 1
  480/1628: End phase -> Roman movement -> 1
   481/1627: MacroCommand(Move [-1,6] to [0,5],Move [0,6] to [1,5]) -> Roman battle -> 3
    CHANCE/688: Close Combat from [0,5] to [0,4] -> Roman battle -> 2
     -288/454: Close Combat from [0,5] to [0,4] -> Carthaginian retreat -> 2
      0/0: Retreat [0,4] to [0,3] -> Roman advance after combat -> 0
      288/454: Retreat [0,4] to [1,3] -> Roman advance after combat -> 2
     69/233: Close Combat from [0,5] to [0,4] -> Roman attacker retreat -> 2
      55/156: Retreat [0,5] to [-1,6] -> Roman battle -> 1
      14/77: Retreat [0,5] to [0,6] -> Roman battle -> 1
    CHANCE/856: Close Combat from [0,5] to [1,4] -> Roman battle -> 5
     53/196: Close Combat from [0,5] to [1,4] -> Roman attacker retreat -> 2
      53/150: Retreat [0,5] to [-1,6] -> Roman battle -> 1
      0/46: Retreat [0,5] to [0,6] -> Roman battle -> 1
     -126/234: Close Combat from [0,5] to [1,4] -> Carthaginian retreat -> 2
      0/0: Retreat [1,4] to [2,3] -> Roman advance after combat -> 0
      126/234: Retreat [1,4] to [1,3] -> Roman advance after combat -> 2
     -130/177: Close Combat from [0,5] to [1,4] -> Roman attacker retreat -> 2
      -63/84: Retreat [0,5] to [-1,6] -> Roman battle -> 1
      -67/93: Retreat [0,5] to [0,6] -> Roman battle -> 1
     54/106: Close Combat from [0,5] to [1,4] -> Carthaginian retreat -> 2
      -54/106: Retreat [1,4] to [2,3] -> Roman advance after combat -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman advance after combat -> 0
     -104/142: Close Combat from [0,5] to [1,4] -> Carthaginian retreat -> 2
      104/142: Retreat [1,4] to [2,3] -> Roman advance after combat -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman advance after combat -> 0
    CHANCE/84: Close Combat from [1,5] to [1,4] -> Roman battle -> 6
     -9/10: Close Combat from [1,5] to [1,4] -> Carthaginian retreat -> 2
      9/10: Retreat [1,4] to [2,3] -> Roman advance after combat -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman advance after combat -> 0
     -18/18: Close Combat from [1,5] to [1,4] -> Carthaginian retreat -> 2
      18/18: Retreat [1,4] to [2,3] -> Roman advance after combat -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman advance after combat -> 0
     -4/7: Close Combat from [1,5] to [1,4] -> Roman attacker retreat -> 2
      -2/5: Retreat [1,5] to [0,6] -> Roman battle -> 2
      -2/2: Retreat [1,5] to [1,6] -> Roman battle -> 2
     -16/25: Close Combat from [1,5] to [1,4] -> Carthaginian retreat -> 2
      16/25: Retreat [1,4] to [2,3] -> Roman advance after combat -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman advance after combat -> 0
     1/8: Close Combat from [1,5] to [1,4] -> Roman attacker retreat -> 2
      1/5: Retreat [1,5] to [0,6] -> Roman battle -> 2
      0/3: Retreat [1,5] to [1,6] -> Roman battle -> 2
     -15/15: Close Combat from [1,5] to [1,4] -> Roman attacker retreat -> 2
      -8/8: Retreat [1,5] to [0,6] -> Roman battle -> 2
      -7/7: Retreat [1,5] to [1,6] -> Roman battle -> 2
 46/372: PlayCard(Order Three Units Left) -> Roman order 3 units -> 1
  46/371: End phase -> Roman movement -> 1
   45/370: MacroCommand(Move [-1,6] to [0,5],Move [0,6] to [1,5]) -> Roman battle -> 3
    CHANCE/275: Close Combat from [0,5] to [1,4] -> Roman battle -> 5
     3/79: Close Combat from [0,5] to [1,4] -> Carthaginian retreat -> 2
      -3/79: Retreat [1,4] to [2,3] -> Roman advance after combat -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman advance after combat -> 0
     2/56: Close Combat from [0,5] to [1,4] -> Roman attacker retreat -> 2
      2/29: Retreat [0,5] to [-1,6] -> Roman battle -> 1
      0/27: Retreat [0,5] to [0,6] -> Roman battle -> 1
     -26/45: Close Combat from [0,5] to [1,4] -> Carthaginian retreat -> 2
      26/45: Retreat [1,4] to [2,3] -> Roman advance after combat -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman advance after combat -> 0
     10/35: Close Combat from [0,5] to [1,4] -> Carthaginian retreat -> 2
      -10/35: Retreat [1,4] to [2,3] -> Roman advance after combat -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman advance after combat -> 0
     -20/59: Close Combat from [0,5] to [1,4] -> Roman attacker retreat -> 2
      -10/27: Retreat [0,5] to [-1,6] -> Roman battle -> 1
      -10/32: Retreat [0,5] to [0,6] -> Roman battle -> 1
    CHANCE/60: Close Combat from [0,5] to [0,4] -> Roman battle -> 2
     2/22: Close Combat from [0,5] to [0,4] -> Roman attacker retreat -> 2
      -1/7: Retreat [0,5] to [-1,6] -> Roman battle -> 1
      3/15: Retreat [0,5] to [0,6] -> Roman battle -> 1
     -30/37: Close Combat from [0,5] to [0,4] -> Carthaginian retreat -> 2
      30/37: Retreat [0,4] to [1,3] -> Roman advance after combat -> 2
      0/0: Retreat [0,4] to [0,3] -> Roman advance after combat -> 0
    CHANCE/36: Close Combat from [1,5] to [1,4] -> Roman battle -> 6
     1/2: Close Combat from [1,5] to [1,4] -> Roman attacker retreat -> 2
      1/2: Retreat [1,5] to [0,6] -> Roman battle -> 2
      0/0: Retreat [1,5] to [1,6] -> Roman battle -> 0
     -10/16: Close Combat from [1,5] to [1,4] -> Carthaginian retreat -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman advance after combat -> 0
      10/16: Retreat [1,4] to [2,3] -> Roman advance after combat -> 2
     -5/5: Close Combat from [1,5] to [1,4] -> Carthaginian retreat -> 2
      5/5: Retreat [1,4] to [2,3] -> Roman advance after combat -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman advance after combat -> 0
     1/1: Close Combat from [1,5] to [1,4] -> Roman attacker retreat -> 2
      1/1: Retreat [1,5] to [0,6] -> Roman battle -> 0
      0/0: Retreat [1,5] to [1,6] -> Roman battle -> 0
     -5/5: Close Combat from [1,5] to [1,4] -> Carthaginian retreat -> 2
      5/5: Retreat [1,4] to [2,3] -> Roman advance after combat -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman advance after combat -> 0
     -6/6: Close Combat from [1,5] to [1,4] -> Roman attacker retreat -> 2
      -3/3: Retreat [1,5] to [0,6] -> Roman battle -> 2
      -3/3: Retreat [1,5] to [1,6] -> Roman battle -> 2
"
`;

exports[`MCTS snapshot testing close combat from light to heavy 1`] = `
"-19/30: undefined -> Roman play one card -> 2
 -10/19: PlayCard(Order Three Units Left) -> Roman order 3 units -> 1
  -10/18: End phase -> Roman movement -> 1
   -11/17: MacroCommand(Move [2,3] to [3,2]) -> Roman battle -> 1
    CHANCE/16: Close Combat from [3,2] to [2,2] -> Roman battle -> 5
     -6/6: Close Combat from [3,2] to [2,2] -> Roman attacker retreat -> 3
      -2/2: Retreat [3,2] to [3,4] -> Roman battle -> 1
       1/1: End phase -> Carthaginian play one card -> 0
      -2/2: Retreat [3,2] to [2,4] -> Roman battle -> 1
       2/2: End phase -> Carthaginian play one card -> 1
      -2/2: Retreat [3,2] to [1,4] -> Roman battle -> 1
       2/2: End phase -> Carthaginian play one card -> 1
     -1/2: Close Combat from [3,2] to [2,2] -> Carthaginian retreat -> 2
      1/2: Retreat [2,2] to [2,1] -> Roman advance after combat -> 2
       1/1: Momentum advance [3,2] to [2,2] -> Roman battle -> 0
       0/0: Skip Momentum Advance in [3,2] -> Roman battle -> 0
      0/0: Retreat [2,2] to [3,1] -> Roman advance after combat -> 0
     2/4: Close Combat from [3,2] to [2,2] -> Carthaginian retreat -> 2
      -2/4: Retreat [2,2] to [3,1] -> Roman advance after combat -> 2
       -1/1: Momentum advance [3,2] to [2,2] -> Roman battle -> 0
       0/2: Skip Momentum Advance in [3,2] -> Roman battle -> 1
      0/0: Retreat [2,2] to [2,1] -> Roman advance after combat -> 0
     -2/2: Close Combat from [3,2] to [2,2] -> Roman attacker retreat -> 3
      -2/2: Retreat [3,2] to [1,4] -> Roman battle -> 1
       1/1: End phase -> Carthaginian play one card -> 0
      0/0: Retreat [3,2] to [3,4] -> Roman battle -> 0
      0/0: Retreat [3,2] to [2,4] -> Roman battle -> 0
     -1/1: Close Combat from [3,2] to [2,2] -> Roman battle -> 1
      1/1: End phase -> Carthaginian play one card -> 0
 -9/11: PlayCard(Order Light Troops) -> Roman order 1 units -> 1
  -9/11: End phase -> Roman movement -> 1
   -10/10: MacroCommand(Move [2,3] to [3,2]) -> Roman battle -> 1
    CHANCE/9: Close Combat from [3,2] to [2,2] -> Roman battle -> 4
     2/2: Close Combat from [3,2] to [2,2] -> Carthaginian retreat -> 2
      -2/2: Retreat [2,2] to [3,1] -> Roman advance after combat -> 2
       -1/1: Momentum advance [3,2] to [2,2] -> Roman battle -> 0
       0/0: Skip Momentum Advance in [3,2] -> Roman battle -> 0
      0/0: Retreat [2,2] to [2,1] -> Roman advance after combat -> 0
     -4/4: Close Combat from [3,2] to [2,2] -> Roman attacker retreat -> 3
      -2/2: Retreat [3,2] to [1,4] -> Roman battle -> 1
       1/1: End phase -> Carthaginian play one card -> 0
      -1/1: Retreat [3,2] to [3,4] -> Roman battle -> 1
       1/1: End phase -> Carthaginian play one card -> 0
      -1/1: Retreat [3,2] to [2,4] -> Roman battle -> 1
       1/1: End phase -> Carthaginian play one card -> 0
     -1/1: Close Combat from [3,2] to [2,2] -> Roman battle -> 1
      1/1: End phase -> Carthaginian play one card -> 0
     -1/1: Close Combat from [3,2] to [2,2] -> Roman attacker retreat -> 3
      0/0: Retreat [3,2] to [1,4] -> Roman battle -> 0
      0/0: Retreat [3,2] to [2,4] -> Roman battle -> 0
      -1/1: Retreat [3,2] to [3,4] -> Roman battle -> 0
"
`;

exports[`MCTS snapshot testing melee 1`] = `
"-98/2000: undefined -> Roman play one card -> 2
 -61/204: PlayCard(Order Three Units Left) -> Roman order 3 units -> 1
  -60/203: End phase -> Roman movement -> 1
   -61/202: MacroCommand(Move [1,5] to [2,4],Move [0,5] to [1,4]) -> Roman battle -> 3
    CHANCE/167: Close Combat from [2,4] to [3,3] -> Roman battle -> 2
     83/97: Close Combat from [2,4] to [3,3] -> Carthaginian retreat -> 3
      0/0: Retreat [3,3] to [4,2] -> Roman advance after combat -> 0
      CHANCE/97: Ignore Flag and Battle Back from [3,3] to [2,4] -> Carthaginian retreat -> 2
      0/1: Retreat [3,3] to [3,2] -> Roman advance after combat -> 0
     6/69: Close Combat from [2,4] to [3,3] -> Roman attacker retreat -> 2
      2/31: Retreat [2,4] to [2,4] -> Roman battle -> 1
      4/38: Retreat [2,4] to [1,5] -> Roman battle -> 1
    CHANCE/27: Close Combat from [2,4] to [2,3] -> Roman battle -> 2
     4/9: Close Combat from [2,4] to [2,3] -> Roman attacker retreat -> 2
      3/6: Retreat [2,4] to [2,4] -> Roman battle -> 1
      1/3: Retreat [2,4] to [1,5] -> Roman battle -> 1
     -6/17: Close Combat from [2,4] to [2,3] -> Carthaginian retreat -> 2
      6/17: Retreat [2,3] to [2,2] -> Roman advance after combat -> 2
      0/0: Retreat [2,3] to [3,2] -> Roman advance after combat -> 0
    CHANCE/9: Close Combat from [1,4] to [2,3] -> Roman battle -> 2
     0/2: Close Combat from [1,4] to [2,3] -> Roman attacker retreat -> 2
      0/2: Retreat [1,4] to [0,5] -> Roman battle -> 2
      0/0: Retreat [1,4] to [1,5] -> Roman battle -> 0
     -6/6: Close Combat from [1,4] to [2,3] -> Carthaginian retreat -> 2
      6/6: Retreat [2,3] to [2,2] -> Roman advance after combat -> 2
      0/0: Retreat [2,3] to [3,2] -> Roman advance after combat -> 0
 -37/1796: PlayCard(Order Heavy Troops) -> Roman order 3 units -> 4
  -29/274: OrderUnit([0,5],[1,5],[3,5]) -> Roman order 3 units -> 1
   -29/274: End phase -> Roman movement -> 1
    -28/273: MacroCommand(Move [1,5] to [2,4],Move [0,5] to [1,4],Move [3,5] to [3,4]) -> Roman battle -> 5
     CHANCE/33: Close Combat from [3,4] to [3,3] -> Roman battle -> 3
      3/7: Close Combat from [3,4] to [3,3] -> Roman attacker retreat -> 2
      -8/22: Close Combat from [3,4] to [3,3] -> Carthaginian retreat -> 3
      1/3: Close Combat from [3,4] to [3,3] -> Roman attacker retreat -> 2
     CHANCE/22: Close Combat from [1,4] to [2,3] -> Roman battle -> 2
      -13/15: Close Combat from [1,4] to [2,3] -> Carthaginian retreat -> 2
      6/6: Close Combat from [1,4] to [2,3] -> Roman attacker retreat -> 2
     CHANCE/18: Close Combat from [2,4] to [2,3] -> Roman battle -> 2
      2/6: Close Combat from [2,4] to [2,3] -> Roman attacker retreat -> 2
      -9/11: Close Combat from [2,4] to [2,3] -> Carthaginian retreat -> 2
     CHANCE/194: Close Combat from [2,4] to [3,3] -> Roman battle -> 2
      5/83: Close Combat from [2,4] to [3,3] -> Roman attacker retreat -> 2
      77/110: Close Combat from [2,4] to [3,3] -> Carthaginian retreat -> 3
     CHANCE/9: Close Combat from [3,4] to [4,3] -> Roman battle -> 6
      -2/2: Close Combat from [3,4] to [4,3] -> Carthaginian retreat -> 2
      -2/2: Close Combat from [3,4] to [4,3] -> Carthaginian retreat -> 2
      -1/1: Close Combat from [3,4] to [4,3] -> Roman attacker retreat -> 2
      -1/1: Close Combat from [3,4] to [4,3] -> Carthaginian retreat -> 2
      -1/1: Close Combat from [3,4] to [4,3] -> Roman attacker retreat -> 2
      -1/1: Close Combat from [3,4] to [4,3] -> Roman attacker retreat -> 2
  -25/325: OrderUnit([0,5],[1,5],[2,5]) -> Roman order 3 units -> 1
   -25/324: End phase -> Roman movement -> 1
    -24/323: MacroCommand(Move [1,5] to [2,4],Move [2,5] to [3,4],Move [0,5] to [1,4]) -> Roman battle -> 5
     CHANCE/11: Close Combat from [3,4] to [4,3] -> Roman battle -> 5
      0/0: Close Combat from [3,4] to [4,3] -> Roman attacker retreat -> 0
      -3/3: Close Combat from [3,4] to [4,3] -> Carthaginian retreat -> 2
      -2/2: Close Combat from [3,4] to [4,3] -> Carthaginian retreat -> 2
      -2/2: Close Combat from [3,4] to [4,3] -> Carthaginian retreat -> 2
      1/3: Close Combat from [3,4] to [4,3] -> Roman attacker retreat -> 2
     CHANCE/66: Close Combat from [3,4] to [3,3] -> Roman battle -> 2
      17/23: Close Combat from [3,4] to [3,3] -> Roman attacker retreat -> 2
      8/42: Close Combat from [3,4] to [3,3] -> Carthaginian retreat -> 3
     CHANCE/23: Close Combat from [2,4] to [2,3] -> Roman battle -> 2
      5/7: Close Combat from [2,4] to [2,3] -> Roman attacker retreat -> 3
      -6/15: Close Combat from [2,4] to [2,3] -> Carthaginian retreat -> 2
     CHANCE/213: Close Combat from [2,4] to [3,3] -> Roman battle -> 2
      102/131: Close Combat from [2,4] to [3,3] -> Carthaginian retreat -> 3
      42/81: Close Combat from [2,4] to [3,3] -> Roman attacker retreat -> 3
     CHANCE/13: Close Combat from [1,4] to [2,3] -> Roman battle -> 2
      -6/9: Close Combat from [1,4] to [2,3] -> Carthaginian retreat -> 2
      1/3: Close Combat from [1,4] to [2,3] -> Roman attacker retreat -> 2
  1/561: OrderUnit([1,5],[2,5],[3,5]) -> Roman order 3 units -> 1
   1/561: End phase -> Roman movement -> 1
    2/560: MacroCommand(Move [1,5] to [2,4],Move [2,5] to [3,4],Move [3,5] to [4,4]) -> Roman battle -> 5
     CHANCE/62: Close Combat from [3,4] to [3,3] -> Roman battle -> 2
      7/25: Close Combat from [3,4] to [3,3] -> Roman attacker retreat -> 3
      0/36: Close Combat from [3,4] to [3,3] -> Carthaginian retreat -> 3
     CHANCE/57: Close Combat from [3,4] to [4,3] -> Roman battle -> 5
      -5/7: Close Combat from [3,4] to [4,3] -> Carthaginian retreat -> 2
      -2/10: Close Combat from [3,4] to [4,3] -> Roman attacker retreat -> 3
      -2/7: Close Combat from [3,4] to [4,3] -> Carthaginian retreat -> 2
      2/20: Close Combat from [3,4] to [4,3] -> Carthaginian retreat -> 2
      12/12: Close Combat from [3,4] to [4,3] -> Roman attacker retreat -> 3
     CHANCE/23: Close Combat from [2,4] to [2,3] -> Roman battle -> 2
      -2/8: Close Combat from [2,4] to [2,3] -> Roman attacker retreat -> 2
      -6/14: Close Combat from [2,4] to [2,3] -> Carthaginian retreat -> 2
     CHANCE/19: Close Combat from [4,4] to [4,3] -> Roman battle -> 5
      2/2: Close Combat from [4,4] to [4,3] -> Roman attacker retreat -> 2
      -5/5: Close Combat from [4,4] to [4,3] -> Carthaginian retreat -> 2
      -1/3: Close Combat from [4,4] to [4,3] -> Carthaginian retreat -> 2
      0/4: Close Combat from [4,4] to [4,3] -> Roman attacker retreat -> 2
      -4/4: Close Combat from [4,4] to [4,3] -> Carthaginian retreat -> 2
     CHANCE/402: Close Combat from [2,4] to [3,3] -> Roman battle -> 2
      127/237: Close Combat from [2,4] to [3,3] -> Carthaginian retreat -> 3
      89/164: Close Combat from [2,4] to [3,3] -> Roman attacker retreat -> 2
  16/636: OrderUnit([0,5],[2,5],[3,5]) -> Roman order 3 units -> 1
   16/636: End phase -> Roman movement -> 1
    15/635: MacroCommand(Move [2,5] to [2,4],Move [0,5] to [1,4],Move [3,5] to [3,4]) -> Roman battle -> 5
     CHANCE/14: Close Combat from [1,4] to [2,3] -> Roman battle -> 2
      2/3: Close Combat from [1,4] to [2,3] -> Roman attacker retreat -> 2
      -10/10: Close Combat from [1,4] to [2,3] -> Carthaginian retreat -> 2
     CHANCE/11: Close Combat from [3,4] to [4,3] -> Roman battle -> 6
      -1/3: Close Combat from [3,4] to [4,3] -> Roman attacker retreat -> 2
      -2/2: Close Combat from [3,4] to [4,3] -> Carthaginian retreat -> 2
      -2/2: Close Combat from [3,4] to [4,3] -> Carthaginian retreat -> 2
      -1/1: Close Combat from [3,4] to [4,3] -> Carthaginian retreat -> 2
      -1/1: Close Combat from [3,4] to [4,3] -> Roman attacker retreat -> 2
      -1/1: Close Combat from [3,4] to [4,3] -> Roman attacker retreat -> 2
     CHANCE/80: Close Combat from [2,4] to [3,3] -> Roman battle -> 2
      -4/47: Close Combat from [2,4] to [3,3] -> Carthaginian retreat -> 3
      26/32: Close Combat from [2,4] to [3,3] -> Roman attacker retreat -> 2
     CHANCE/517: Close Combat from [3,4] to [3,3] -> Roman battle -> 3
      189/306: Close Combat from [3,4] to [3,3] -> Carthaginian retreat -> 3
      83/127: Close Combat from [3,4] to [3,3] -> Roman attacker retreat -> 2
      72/83: Close Combat from [3,4] to [3,3] -> Roman attacker retreat -> 2
     CHANCE/16: Close Combat from [2,4] to [2,3] -> Roman battle -> 2
      -6/9: Close Combat from [2,4] to [2,3] -> Carthaginian retreat -> 2
      -1/6: Close Combat from [2,4] to [2,3] -> Roman attacker retreat -> 2
"
`;
